File: 2024/src/main/java/dpr/aoc2024/Day03.java
package dpr.aoc2024;

import dpr.commons.Day;
import dpr.commons.Util;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Day03 implements Day {
    @Override
    public void execute() {
        var lines = Util.getNotEmptyLinesFromFile(dayNum(), "input.txt");
        System.out.println(part1(lines));
        System.out.println(part2(lines));
    }

    @Override
    public int dayNum() {
        return 3;
    }

    Object part1(List<String> lines) {
        String program = String.join("", lines);
        return calculateSum(program);
    }

    private static long calculateSum(String program) {
        Pattern pattern = Pattern.compile("mul\\((\\d{1,3}),(\\d{1,3})\\)");
        Matcher matcher = pattern.matcher(program);
        var sum = 0L;
        while (matcher.find()) {
            long left = Integer.parseInt(matcher.group(1));
            long right = Integer.parseInt(matcher.group(2));
            sum += left * right;
        }
        return sum;
    }

    Object part2(List<String> lines) {
        String program = String.join("", lines);
        String[] split = program.split("((?=do(n't|)\\(\\))|(?<=do(n't|)\\(\\)))");
        boolean enabled = true;
        long sum = 0L;
        for (String cur : split) {
            if (cur.equals("do()")) {
                enabled = true;
            } else if (cur.equals("don't()")) {
                enabled = false;
            } else if (enabled) {
                sum += calculateSum(cur);
            }
        }
        return sum;
        //115739063 is too high
    }
}

File: 2020/src/main/kotlin/dpr/aoc2020/Day03.kt
package dpr.aoc2020

import dpr.commons.Util

object Day03 {
    @JvmStatic
    fun main(args: Array<String>) = Util.measureTime {
//        val input = Util.getFileContent("/03/test1")
        val input = Util.getNotEmptyLinesFromFile("/03/input.txt")
        println(part1(input))
        println(part2(input))
    }

    private fun part1(input: List<String>): Any {
        val stepX = 3
        val stepY = 1
        return countTrees(input, stepY, stepX)
    }

    private fun countTrees(lines: List<String>, stepY: Int, stepX: Int): Long {
        var i = 0
        var j = 0
        var trees = 0L
        while (i < lines.size) {
            trees += if (lines[i][j % lines[0].length] == '#') 1 else 0
            i += stepY
            j += stepX
        }
        return trees
    }

    private fun part2(input: List<String>): Any {
        return countTrees(input, 1, 1) *
            countTrees(input, 1, 3) *
            countTrees(input, 1, 5) *
            countTrees(input, 1, 7) *
            countTrees(input, 2, 1)

    }

}

